# A pipeline used to identify Ks peaks by fitting Gaussian Mixture Model (GMM)
This pipline aimed to estimate the Ks peaks corresponding to WGD events of different ages by fitting Gaussian Mixture Model [(GMM)](https://scikit-learn.org/stable/modules/generated/sklearn.mixture.GaussianMixture.html#sklearn.mixture.GaussianMixture) to Ks distributions of intra-species syntenic blocks. A visualization tool was incorporated in this pipeline to show the Ks distribution and the curve fitted by GMM. It depends on [MCScanX](http://chibba.pgml.uga.edu/mcscan2/) and [calculate_Ka_Ks_pipeline](https://github.com/qiao-xin/Scripts_for_GB/tree/master/calculate_Ka_Ks_pipeline).

| | |
| --- | --- |
| Authors | Xin Qiao ([qiaoxin](https://github.com/qiao-xin)) |
| | Qionghou Li ([Qionghou Li](https://github.com/LQHHHHH)) |
| Email   | <qiaoxinqx2011@126.com> |

## Dependencies

- [Perl](https://www.perl.org)
- [BioPerl](https://bioperl.org)
- [Python 2.7](https://www.python.org/)
  - [pandas](http://pandas.pydata.org/)
  - [scipy](https://www.scipy.org/)
  - [scikit-learn](https://scikit-learn.org/stable/index.html)

## Installation

```bash
git clone https://github.com/qiao-xin/Scripts_for_GB.git
```

## Running

Here, we used this pipeline to infer the Ks peaks corresponding to WGD events of different ages in *Arabidopsis thaliana* genome.

### 1. Input Files Preparation

Before running this pipline, some files should be prepared:
- ```Ath.collinearity```: an output file from running [MCScanX](http://chibba.pgml.uga.edu/mcscan2/).
- ```Ath.wgd.kaks```: an output file from running [calculate_Ka_Ks_pipeline](https://github.com/qiao-xin/Scripts_for_GB/tree/master/calculate_Ka_Ks_pipeline).

### 2. Adding Ka, Ks and Ka/Ks into collinearity file 

A typical command could look like this:
~~~bash
perl add_ka_ks_to_collinearity_file.pl Ath
~~~
This command will automatically add Ka, Ks, Ka/Ks values into Ath.collinearity by using Ath.wgd.kaks as input, and produce one output file: ```Ath.collinearity.kaks```.

```
############### Parameters ###############
# MATCH_SCORE: 50
# MATCH_SIZE: 5
# GAP_PENALTY: -1
# OVERLAP_WINDOW: 5
# E_VALUE: 1e-05
# MAX GAPS: 25
############### Statistics ###############
# Number of collinear genes: 7544, Percentage: 27.52
# Number of all genes: 27416
##########################################
## Alignment 0: score=8970.0 e_value=0 N=190 Ath-Chr1&Ath-Chr1 plus Ka Ks Ka/Ks
  0-  0:	AT1G17240.1	AT1G72300.1	      0	0.136575	0.717908	0.19024
  0-  1:	AT1G17290.1	AT1G72330.3	      0	0.0746912	0.749983	0.0995905
  0-  2:	AT1G17310.1	AT1G72350.1	  2e-49	0.349746	1.3647	0.25628
  0-  3:	AT1G17350.2	AT1G72420.1	 5e-132	0.121138	0.9046	0.133913
  0-  4:	AT1G17380.1	AT1G72450.1	  2e-76	0.278683	1.69693	0.164228
  0-  5:	AT1G17400.1	AT1G72490.1	 3e-101	0.176897	1.22428	0.14449
  0-  6:	AT1G17420.1	AT1G72520.1	      0	0.0914749	1.3165	0.0694835
  0-  7:	AT1G17430.1	AT1G72620.1	      0	0.125315	0.760592	0.16476
```

### 3. Calculating average Ks value for each syntenic block

A typical command could look like this:
~~~bash
perl compute_ks_for_synteny_blocks.pl Ath.collinearity.kaks
~~~
This will produce one output file: ```Ath.synteny.blocks.ks.info```, which contains average Ks values for gene pairs contained in each syntenic block.

The data in Ath.synteny.blocks.ks.info looks like this (tab separated):
```
Blocks ID	Location	Block Size	Average Ks	e-value	Score	Orientation
Alignment166	Ath-Chr3&Ath-Chr5	7	1.42200142857143	4.6e-12	330.0	plus
Alignment68	Ath-Chr1&Ath-Chr4	8	3.6449275	6.9e-15	341.0	minus
Alignment140	Ath-Chr2&Ath-ChrM	13	0.329314675	2.6e-36	610.0	plus
Alignment169	Ath-Chr3&Ath-Chr5	40	1.267692325	4.6e-136	1867.0	minus
Alignment121	Ath-Chr2&Ath-Chr4	7	1.48141166666667	4.4e-11	305.0	minus
```

### 4. Estimating Ks peaks from Ks distribution

**USAGE:**
~~~bash
python plot_syntenic_blocks_ks_distri.py Species_Abbr.synteny.blocks.ks.info Components Species_Abbr
~~~
The file ```Species_Abbr.synteny.blocks.ks.info``` was generated by Step 3, for example, "Ath.synteny.blocks.ks.info". The parameter ```Components``` indicates the number of the mixture components, which represent the number of Ks peak. The parameter ```Species_Abbr``` indicates scientific name abbreviation you defined for the investigated species, for example, "Ath".

A typical command could look like this:
~~~bash
python plot_syntenic_blocks_ks_distri.py Ath.synteny.blocks.ks.info 2 Ath
~~~
This command will produce:
- Ks peak values, the related information will be printed when running this command.
- Ath.synteny.blocks.ks.distri.pdf

![Ath.synteny.blocks.ks.distri](/identify_Ks_peaks_by_fitting_GMM/data/Ath.synteny.blocks.ks.distri.png)
*Figure 1: The distribution of Ks values of syntenic blocks within Arabidopsis genome.*

Ks peaks corresponding to WGD events of different ages were inferred by fitting Gaussian mixture models to Ks distribution. 

## Citation
*[Qiao X, Li Q, Yin H, Qi K, Li L, Wang R, Zhang S, Paterson AH: Gene duplication and evolution in recurring polyploidizationâ€“diploidization cycles in plants. Genome Biology 2019, 20:38.](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1650-2)*
